contracts/src/libraries/PlonkVerifier.sol

```diff
@@ -0,0 +1,856 @@
+        uint256[] commScalars; // 0x80
+        // bases of poly comm for MSM
+        BN254.G1Point[] commBases; // 0xa0
+        // proof of evaluations at point `eval_point`
+        BN254.G1Point openingProof; // 0xc0
+        // proof of evaluations at point `next_eval_point`
+        BN254.G1Point shiftedOpeningProof; // 0xe0
+    }
+
+    /// @dev Plonk IOP verifier challenges.
+    struct Challenges {
+        uint256 alpha; // 0x00
+        uint256 alpha2; // 0x20
+        uint256 alpha3; // 0x40
+        uint256 beta; // 0x60
+        uint256 gamma; // 0x80
+        uint256 zeta; // 0xA0
+        uint256 v; // 0xC0
+        uint256 u; // 0xE0
+    }
+
+    /// @dev Batch verify multiple TurboPlonk proofs.
+    /// @param verifyingKeys An array of verifier keys
+    /// @param publicInputs A two-dimensional array of public inputs.
+    /// @param proofs An array of Plonk proofs
+    /// @param extraTranscriptInitMsgs An array of bytes from
+    /// transcript initialization messages
+    function batchVerify(
+        IPlonkVerifier.VerifyingKey[] memory verifyingKeys,
+        uint256[][] memory publicInputs,
+        IPlonkVerifier.PlonkProof[] memory proofs,
+        bytes[] memory extraTranscriptInitMsgs
+    ) external view returns (bool) {
+        if (
+            verifyingKeys.length != proofs.length || publicInputs.length != proofs.length
+                || extraTranscriptInitMsgs.length != proofs.length || proofs.length == 0
+        ) {
+            revert InvalidPlonkArgs();
+        }
+
+        PcsInfo[] memory pcsInfos = new PcsInfo[](proofs.length);
+        for (uint256 i = 0; i < proofs.length; i++) {
+            // validate proofs are proper group/field elements
+            _validateProof(proofs[i]);
+            // validate public input are all proper scalar fields
+            for (uint256 j = 0; j < publicInputs[i].length; j++) {
+                BN254.validateScalarField(publicInputs[i][j]);
+            }
+            // prepare pcs info
+            pcsInfos[i] = _preparePcsInfo(
+                verifyingKeys[i], publicInputs[i], proofs[i], extraTranscriptInitMsgs[i]
+            );
+        }
+
+        return _batchVerifyOpeningProofs(pcsInfos);
+    }
+
+    /// @dev Validate all group points and scalar fields. Revert if
+    /// any are invalid.
+    /// @param proof A Plonk proof
+    function _validateProof(IPlonkVerifier.PlonkProof memory proof) internal pure {
+        BN254.validateG1Point(proof.wire0);
+        BN254.validateG1Point(proof.wire1);
+        BN254.validateG1Point(proof.wire2);
+        BN254.validateG1Point(proof.wire3);
+        BN254.validateG1Point(proof.wire4);
+        BN254.validateG1Point(proof.prodPerm);
+        BN254.validateG1Point(proof.split0);
+        BN254.validateG1Point(proof.split1);
+        BN254.validateG1Point(proof.split2);
+        BN254.validateG1Point(proof.split3);
+        BN254.validateG1Point(proof.split4);
+        BN254.validateG1Point(proof.zeta);
```

Comment:
Is zetaOmega missing?

Reply:
very good catch! G... I failed miserably by not catching this in tests, but you catch it using bare eyes. LOL irony.

Reply:
ah! I know why my test is missing it. 
1. I forget to test for `zetaOmega`
2. `vm.expectRevert()` should only be invoked once in each tests... I misuse it.

Reply:
fixed in [1d7c6c6](https://github.com/EspressoSystems/espresso-sequencer/pull/722/commits/1d7c6c646e93c2077328e1a30d50e75f1327e04f)